syntax = "proto3";

package zecale_proto;

import "google/protobuf/empty.proto";

import "api/snark_messages.proto";

service Aggregator {
    // Fetch the verification key corresponding to the aggregator statement
    // (the statement including multiple calls to the SNARK verification
    // routine). The Zecale verifier contract must be instantiated with this
    // key in order to verify batches on-chain.
    rpc GetVerificationKey(google.protobuf.Empty) returns (zeth_proto.VerificationKey) {}

    // Register a new application. That is, an application name and a
    // corresponding verfication key. After registration, transactions for the
    // app can be submitted.
    //
    // (May not need an endpoint for this depending on the configuration /
    // use-case. For now, this is convenient for testing).
    rpc RegisterApplication(ApplicationRegistration) returns (google.protobuf.Empty) {}

    // Request a proof that verifies a batch of proofs for the given
    // application name.
    //
    // (The only argument of this function is the name of the application for
    // which we want to generate an aggregate proof. The proofs to batch should
    // have already been deposited in the aggregator tx pool. Returns the proof
    // of CI for the validity of the batch of proofs. This endpoint won't
    // necessarily be useful in practice, but this is useful for some manual
    // triggering for now.)
    rpc GenerateAggregateProof(ApplicationName) returns (zeth_proto.ExtendedProof) {}

    // Function to submit a transaction to aggregate
    rpc SubmitTransaction(TransactionToAggregate) returns (google.protobuf.Empty) {}
}

message ApplicationName {
    string name = 1;
}

message ApplicationRegistration {
    string name = 1;
    zeth_proto.VerificationKey vk = 2;
}

// A transaction for a specific application (determined by `application_name`),
// to be later aggregated into a batch.
message TransactionToAggregate {
    string application_name = 1;
    zeth_proto.ExtendedProof extended_proof = 2;
    // Only if an incentive structure is in place and fees are supported
    int32 fee_in_wei = 3;
}
